README.md
# NAME

MySQL::GrantPrivileges - Generates GRANT statements by loading data from mysqld.

# SYNOPSIS

    use MySQL::GrantPrivileges;

    my $priv = Mysql::GrantPrivileges->new(
        host     => '127.0.0.1',
        port     => '3306',
        user     => 'root',
        password => 'secret',
    );

    # outputs SQL
    print $priv->to_string;

# DESCRIPTION

MySQL::GrantPrivileges is a generator of GRANT statements by loading data from
mysqld. This module accesses mysql.user and mysql.db to generate statements.

This distribution includes a script named [grant-privileges.pl](http://search.cpan.org/perldoc?grant-privileges.pl). So you can
use this script to generate GRANT statements.

# LICENSE

Copyright (C) 2013 Yoshihiro Sasaki

This is free software licensed under

    The BSD 2-Clause License

The full text of the license can be found in the LICENSE file included with
this distribution.

# AUTHOR

Yoshihiro Sasaki <ysasaki@cpan.org>


Changes
Revision history for Perl extension MySQL-GrantPrivileges

{{$NEXT}}

0.02 2013-08-05T09:13:19Z
    - Fixes a issue that parsing short options.
    - Adds a feature that shows password prompt.

0.01 2013-08-05T03:01:59Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'MySQL-GrantPrivileges',
    module_name     => 'MySQL::GrantPrivileges',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/MySQL/GrantPrivileges.pm
package MySQL::GrantPrivileges;

use 5.008005;
use Moo;
use DBI;

our $VERSION = "0.02";

has 'user' => (
    is      => 'ro',
    default => 'root'
);

has 'password' => (
    is      => 'ro',
    default => '',
);

has 'host' => (
    is      => 'ro',
    default => '127.0.0.1',
);

has 'port' => (
    is      => 'ro',
    default => '3306',
);

has 'dsn' => ( is => 'lazy' );

has 'dbh' => ( is => 'lazy', predicate => 1 );

has 'privileges' => ( is => 'lazy' );

# _build_*

sub _build_dsn {
    my $self = shift;
    return sprintf 'dbi:mysql:host=%s;port=%s', $self->host, $self->port;
}

sub _build_dbh {
    my $self = shift;
    my $dbh
        = DBI->connect( $self->dsn, $self->user, $self->password,
        { RaiseError => 1, AutoCommit => 1 } )
        or Carp::croak( "Cannot connect database: " . DBI->errstr );

    return $dbh;
}

# List of column names that don't match with privilege names.
# http://dev.mysql.com/doc/refman/5.6/en/privileges-provided.html
my %EXCEPTION_COLUMNS = (
    'grant_priv'            => 'GRANT OPTION',
    'create_tmp_table_priv' => 'CREATE TEMPORARY TABLES',
    'proxies_priv'          => 'PROXY',
    'repl_client_priv'      => 'REPLICATION CLIENT',
    'repl_slave_priv'       => 'REPLICATION SLAVE',
    'show_db_priv'          => 'SHOW DATABASES',
);

sub _build_privileges {
    my $self = shift;
    my $dbh  = $self->dbh;

    my $privileges = {};
    for (qw(user db)) {
        my $fetched
            = $dbh->selectall_arrayref(
            sprintf( "SELECT * FROM mysql.%s", $_ ),
            { Slice => {} } );

        if ( $dbh->err and ( my $errstr = $dbh->errstr ) ) {
            $dbh->disconnect;
            die "Cannot fetch privileges from database: $errstr";
        }

        my @filtered;
        for my $datum (@$fetched) {
            my $row = {};

            for (qw(host db user password)) {
                $row->{$_} = $datum->{ ucfirst($_) }
                    if defined $datum->{ ucfirst($_) };
            }

            $row->{privileges} = [];
            for my $key ( grep /^\w+_priv$/, keys %$datum ) {
                if ( $datum->{$key} eq 'Y' ) {
                    my $priv;
                    if ( $EXCEPTION_COLUMNS{ lc $key } ) {
                        $priv = $EXCEPTION_COLUMNS{ lc $key };
                    }
                    else {
                        ($priv) = $key =~ m/^(\w+)_priv$/;
                        $priv =~ s/_/ /g;
                    }
                    push @{ $row->{privileges} }, uc $priv;
                }
            }
            push @filtered, $row;
        }

        $privileges->{$_} = \@filtered;
    }

    return $privileges;
}

# method

sub to_string {
    my $self       = shift;
    my $privileges = $self->privileges;

    my @line;
    for my $type (qw(user db)) {

        for my $row ( @{ $privileges->{$type} } ) {

            my $privs
                = scalar @{ $row->{privileges} }
                ? join ', ', @{ $row->{privileges} }
                : 'USAGE';

            my $db_and_tbl = '*.*';
            if ( $row->{db} ) {
                $db_and_tbl = sprintf '`%s`.*', $row->{db};
            }

            my $sql = sprintf( "GRANT %s ON %s TO '%s'\@'%s'",
                $privs, $db_and_tbl, $row->{user}, $row->{host} );

            if ( $row->{password} ) {
                $sql .= sprintf " IDENTIFIED BY PASSWORD '%s'",
                    $row->{password};
            }

            $sql .= ';';

            push @line, $sql;
        }
    }

    return join "\n", @line;
}

sub DEMOLISH {
    my $self = shift;
    $self->dbh->disconnect if $self->has_dbh;
}

1;
__END__

=encoding utf-8

=head1 NAME

MySQL::GrantPrivileges - Generates GRANT statements by loading data from mysqld.

=head1 SYNOPSIS

    use MySQL::GrantPrivileges;

    my $priv = Mysql::GrantPrivileges->new(
        host     => '127.0.0.1',
        port     => '3306',
        user     => 'root',
        password => 'secret',
    );

    # outputs SQL
    print $priv->to_string;

=head1 DESCRIPTION

MySQL::GrantPrivileges is a generator of GRANT statements by loading data from
mysqld. This module accesses mysql.user and mysql.db to generate statements.

This distribution includes a script named L<grant-privileges.pl>. So you can
use this script to generate GRANT statements.

=head1 LICENSE

Copyright (C) 2013 Yoshihiro Sasaki

This is free software licensed under

    The BSD 2-Clause License

The full text of the license can be found in the LICENSE file included with
this distribution.

=head1 AUTHOR

Yoshihiro Sasaki E<lt>ysasaki@cpan.orgE<gt>

=cut

